### PR TITLE
Audio only

### PR DESCRIPTION
**Pull request by krystalmejia24**

### CODE REVIEW

File: provider/bitmovin/bitmovin.go [hunk 1]
```diff
@@ -196,11 +198,9 @@ func (p *bitmovinProvider) Transcode(job *db.Job) (*provider.JobStatus, error) {
 		SelectionMode: model.StreamSelectionMode_AUTO,
 	}
 
-	vidInputStreams := []model.StreamInput{inputStream}
-	audInputStreams := []model.StreamInput{inputStream}
-
+	audioOnly := make([]bool, len(job.Outputs))
```
Comment (zsiec) [inline provider/bitmovin/bitmovin.go]: The loop that follows is checking for HLS outputs configured and the loop bails as soon as it sees one, I think we'll want to check for audio-only things elsewhere as we aren't guaranteed to fill this slice with the correct values.

File: provider/bitmovin/bitmovin.go [hunk 2]
```diff
@@ -260,16 +270,25 @@ func (p *bitmovinProvider) Transcode(job *db.Job) (*provider.JobStatus, error) {
...
 		if err != nil {
 			return nil, err
 		}
 
-		audCfgID, err := configuration.AudCfgIDFrom(details.CustomData)
-		if err != nil {
-			return nil, err
+		audCfgID := ""
+		if audioOnly[i] {
+			audCfg, err := p.api.Encoding.Configurations.Audio.Aac.Get(presetID)
```
Comment (zsiec) [inline provider/bitmovin/bitmovin.go]: I think we need to do a larger refactor as this really couples this code to the underlying details. This assumes we'd want `aac` for any audio only codec. I think this method is already quite large and hard to follow, we need to break it up or consider ways we can abstract things so we're not making changes here.

File: provider/bitmovin/bitmovin.go [hunk 3]
```diff
@@ -260,16 +270,25 @@ func (p *bitmovinProvider) Transcode(job *db.Job) (*provider.JobStatus, error) {
...
+			audCfgID = audCfg.Id
```
Comment (zsiec) [inline provider/bitmovin/bitmovin.go]: Usually, anytime you call a function that returns an error, you should check that error first before using the returned value as there are no guarantee on what it'll return. If it returned a pointer and you try to reference a prop, you may see panics

File: provider/bitmovin/internal/configuration/aac.go [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+import (
+	"github.com/bitmovin/bitmovin-api-sdk-go"
+	"github.com/bitmovin/bitmovin-api-sdk-go/model"
+	"github.com/cbsinteractive/video-transcoding-api/db"
+	"github.com/cbsinteractive/video-transcoding-api/provider/bitmovin/internal/configuration/codec"
+	"github.com/cbsinteractive/video-transcoding-api/provider/bitmovin/internal/types"
+	"github.com/pkg/errors"
+)
+
+// AAC is a configuration service for content in this codec pair
```
Comment (zsiec) [inline provider/bitmovin/internal/configuration/aac.go]: this comment may be misleading as there isn't a pair in this case

File: provider/bitmovin/internal/configuration/aac.go [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+	if err != nil {
+		return "", err
+	}
+
+	return audCfgID, nil
+}
+
+// Get retrieves audio with a presetID
+// the function will return a boolean indicating whether the video
+// configuration was found, a config object and an optional error
```
Comment (zsiec) [inline provider/bitmovin/internal/configuration/aac.go]: We should update this comment as well.

File: provider/bitmovin/internal/configuration/aac.go [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+func (c *AAC) Get(cfgID string) (bool, Details, error) {
+	audCfg, customData, err := c.audioConfigWithCustomDataFrom(cfgID)
+	if err != nil {
+		return false, Details{}, errors.Wrapf(err, "getting the audio configuration with ID %q", cfgID)
+	}
+
+	return true, Details{nil, audCfg, customData}, nil
+}
+
+// Delete removes the audio / video configurations
```
Comment (zsiec) [inline provider/bitmovin/internal/configuration/aac.go]: We should update this comment as well.

File: provider/bitmovin/internal/configuration/h264_aac.go [hunk 7]
```diff
@@ -21,7 +21,7 @@ func NewH264AAC(api *bitmovin.BitmovinApi) *H264AAC {
 
 // Create will create a new H264AAC configuration based on a preset
 func (c *H264AAC) Create(preset db.Preset) (string, error) {
-	audCfgID, err := codec.NewAAC(c.api, preset.Audio.Bitrate)
+	audCfgID, err := codec.NewAAC(c.api, preset.Audio.Bitrate, customDataWith("", preset.Container))
```
Comment (zsiec) [inline provider/bitmovin/internal/configuration/h264_aac.go]: Do we need to send this data in this case? Should we have another constructor that doesn't take custom data if we have cases where we don't use it?

File: provider/bitmovin/internal/configuration/store.go [hunk 8]
```diff
@@ -29,8 +29,8 @@ const (
 )
 
 // ContainerFrom is a helper for extracting the container value from customData
-func ContainerFrom(data types.CustomData) (string, error) {
-	contnr, err := types.CustomDataStringValAtKeys(data, customDataKeyContainer, customDataKeyContainerName)
+func ContainerFrom(details Details) (string, error) {
```
Comment (zsiec) [inline provider/bitmovin/internal/configuration/store.go]: Curious about why we changed the input datatype for this function? Are we using other properties from Details here?

File: provider/bitmovin/internal/container/hls.go [hunk 9]
```diff
@@ -63,28 +63,30 @@ func (a *HLSAssembler) Assemble(cfg AssemblerCfg) error {
...
-		SegmentNaming: "seg_%number%.ts",
-		Streams:       []model.MuxingStream{cfg.VidMuxingStream},
-		Outputs: []model.EncodingOutput{
-			storage.EncodingOutputFrom(cfg.OutputID, path.Join(cfg.ManifestMasterPath, cfg.VidCfgID)),
-		},
-	})
-	if err != nil {
-		return errors.Wrap(err, "creating video ts muxing")
-	}
+	if !(cfg.VidMuxingStream == model.MuxingStream{}) {
```
Comment (zsiec) [inline provider/bitmovin/internal/container/hls.go]: This essentially means if we don't have video, we could be creating the audio HLS muxing/media, but then do nothing with it?

File: provider/bitmovin/bitmovin.go (hunk 1 above)
Comment (krystalmejia24) [reply provider/bitmovin/bitmovin.go]: Should we just create a seperate loop prior for evaluating audio only outputs? 

This will tie into the larger discussion of refactoring this entire Transcode function within bitmovin provider. To your point, its too large.

File: provider/bitmovin/internal/configuration/h264_aac.go (hunk 7 above)
Comment (krystalmejia24) [reply provider/bitmovin/internal/configuration/h264_aac.go]: I only did it this way because I figured we wanted one constructor for all. Its how its done for configs that are video only. 

https://github.com/cbsinteractive/video-transcoding-api/blob/master/provider/bitmovin/internal/configuration/h265.go#L24

is customData something we define ourselves or is that defined within bitmovin api? I know when the transcoding process starts there is a lot of dependencies on this customData struct. we can create a new map interface with the container only being set that we can call for the configs that aren't audio/video pairs.

Comment (krystalmejia24) [reply provider/bitmovin/internal/configuration/h264_aac.go]: "I know when the transcoding process starts there is a lot of dependencies on this customData struct"

when we abstract that transcode function, than we will probably not need this custom data. As i was doing this, i saw it as a dependency so forced myself to do it this way.

Comment (zsiec) [reply provider/bitmovin/internal/configuration/h264_aac.go]: Yeah I think all the video-only configs use the custom data so we only have one constructor for those, we should be able to create different constructors if we run into a case where the values aren't relevant. 

`CustomData` is a Bitmovin thing: https://github.com/bitmovin/bitmovin-api-sdk-go/blob/master/model/custom_data.go and we can easily build a container-only structure similar to how we build this today: https://github.com/cbsinteractive/video-transcoding-api/blob/master/provider/bitmovin/internal/configuration/store.go#L51

File: provider/bitmovin/internal/configuration/store.go (hunk 8 above)
Comment (krystalmejia24) [reply provider/bitmovin/internal/configuration/store.go]: GOOD CATCH. At some point when I was working on this, the thought process was audio only won't have customData so if i passed the details, i could check if this existed. Eventually I ended up initializing customData with only the container present, and a nill audCfgID. So will need to change this back.

File: provider/bitmovin/internal/container/hls.go (hunk 9 above)
Comment (krystalmejia24) [reply provider/bitmovin/internal/container/hls.go]: hmm. I dont think I understand the q. I thought all i was doing here was stopping the vid muxing and media from being created. Looks like all this thing does is create in api, and validates no errors thrown. then the transcode function will start the job. What am I missing?

Comment (zsiec) [reply provider/bitmovin/internal/container/hls.go]: Sorry, I could have been a lot clearer. This method creates all the necessary resources for an HLS stream. It starts by creating an audio `TSMuxing` and an `HLSAudioMedia` resource with the API (unless asked to skip audio). Next the video `TSMuxing` is created and the whole HLS stream gets created with `a.api.HLSStreams.Create`

If we're not calling `a.api.HLSStreams.Create()` if the video description is blank, then we have those audio `TSMuxing` and `HLSAudioMedia` dangling, never attached to anything.

Comment (zsiec) [reply provider/bitmovin/internal/container/hls.go]: @krystalmejia24 I'm wrong! It looks like this logic is correct as these are attached to the provided `manifestID`, ignore me!

**Comment by zsiec** [comment]
closing, changes were merged in https://github.com/cbsinteractive/video-transcoding-api/pull/93 and https://github.com/cbsinteractive/video-transcoding-api/pull/94
